rholang/src/main/scala/coop/rchain/rholang/crypto/Blake2b512Block.scala

```diff
@@ -0,0 +1,152 @@
+      internalState(posB) = rotr64(internalState(posB) ^ internalState(posC), 24);
+      internalState(posA) = internalState(posA) + internalState(posB) + m2;
+      internalState(posD) = rotr64(internalState(posD) ^ internalState(posA), 16);
+      internalState(posC) = internalState(posC) + internalState(posD);
+      internalState(posB) = rotr64(internalState(posB) ^ internalState(posC), 63);
+    }
+
+    Array.copy(chainValue, 0, internalState, 0, CHAIN_VALUE_LENGTH)
+    Array.copy(IV, 0, internalState, CHAIN_VALUE_LENGTH, 4)
+    val f0    = if (peek) 0xffffffffffffffffL else 0x0L
+    val newT0 = t0 + BLOCK_LENGTH_BYTES
+    val newT1 = if (newT0 == 0) t1 + 1 else t1
+    internalState(12) = newT0 ^ IV(4)
+    internalState(13) = newT1 ^ IV(5)
+    internalState(14) = f0 ^ IV(6)
+    internalState(15) = IV(7)
+
+    val m: Array[Long] = new Array(BLOCK_LENGTH_LONGS)
+    0.until(BLOCK_LENGTH_LONGS).foreach { i =>
+      m(i) = Pack.littleEndianToLong(msg, offset + i * 8)
+    }
+
+    0.until(ROUNDS).foreach { round =>
+      // columns
+      g(m(SIGMA(round)(0).toInt), m(SIGMA(round)(1).toInt), 0, 4, 8, 12);
+      g(m(SIGMA(round)(2).toInt), m(SIGMA(round)(3).toInt), 1, 5, 9, 13);
+      g(m(SIGMA(round)(4).toInt), m(SIGMA(round)(5).toInt), 2, 6, 10, 14);
+      g(m(SIGMA(round)(6).toInt), m(SIGMA(round)(7).toInt), 3, 7, 11, 15);
+      // diagonals
+      g(m(SIGMA(round)(8).toInt), m(SIGMA(round)(9).toInt), 0, 5, 10, 15);
+      g(m(SIGMA(round)(10).toInt), m(SIGMA(round)(11).toInt), 1, 6, 11, 12);
+      g(m(SIGMA(round)(12).toInt), m(SIGMA(round)(13).toInt), 2, 7, 8, 13);
+      g(m(SIGMA(round)(14).toInt), m(SIGMA(round)(15).toInt), 3, 4, 9, 14);
+    }
+    if (!peek) {
+      t0 = newT0
+      t1 = newT1
+    }
+    0.until(CHAIN_VALUE_LENGTH).foreach { i =>
+      newChainValue(i) = chainValue(i) ^ internalState(i) ^ internalState(i + 8)
+    }
+  }
+
+}
+
+object Blake2b512Block {
+  def apply(): Blake2b512Block = {
+    val result = new Blake2b512Block
+    Array.copy(IV, 0, result.chainValue, 0, CHAIN_VALUE_LENGTH)
+    result.chainValue(0) ^= PARAM_VALUE
+    result
+  }
+
+  def apply(src: Blake2b512Block): Blake2b512Block = {
+    val result = new Blake2b512Block
+    Array.copy(src.chainValue, 0, result.chainValue, 0, CHAIN_VALUE_LENGTH)
+    result.t0 = src.t0
+    result.t1 = src.t1
+    result
+  }
+
+  implicit val serialize: Serialize[Blake2b512Block] = new Serialize[Blake2b512Block] {
+    override def encode(block: Blake2b512Block) = {
+      val result: Array[Byte] = new Array[Byte](80)
+      Pack.longToLittleEndian(block.chainValue, result, 0)
+      Pack.longToLittleEndian(block.t0, result, DIGEST_LENGTH_BYTES)
+      Pack.longToLittleEndian(block.t1, result, DIGEST_LENGTH_BYTES + 8)
+      result
+    }
+
+    override def decode(bytes: Array[Byte]): Either[Throwable, Blake2b512Block] =
+      if (bytes.length != 80)
+        Left(new Error("Invalid decode of Blake2b512Block"))
```

Comment:
Did you mean "Invalid encoding"?  It's not this function's fault that the given input's length is wrong.